middle_end/flambda/inlining/inlining_report.ml

```diff
@@ -0,0 +1,176 @@
+          code_id : Code_id.t;
+          decision : Inlining_decision.Call_site_decision.t;
+        }
+
+  end
+
+  module At_function_declaration = struct
+
+    type pass =
+      | Before_simplify
+      | After_simplify
+
+    type t = {
+      pass : pass;
+      code_id : Code_id.t;
+      decision : Inlining_decision.Function_declaration_decision.t;
+    }
+
+  end
+
+  type t =
+    | At_call_site of At_call_site.t
+    | At_function_declaration of At_function_declaration.t
+
+end
+
+(* Log for a whole round
+
+   Currently, we make use of the precise ordering of decisions to print
+   the log in a somewhat hierarchical manner. More precisely, we rely on the
+   fact that there are two decisions for a function declarations: before and after
+   simplify. This make it possible to treat those two decisions as "entering"
+   and "exiting" the function's body, considering all decisions that occur in between
+   to pertain to the body of the function, using the order of the list insertions of
+   the {record_decision} function.
+
+   On the other hand, flambda1 had a non-mutable way of storing this information as
+   a closure_stack (see trunk:middle_end/flambda/inlining_stats.ml) in the environment
+   of the simplification process. This could be replicated here, i.e. we could store
+   something akin to flambda1's Closure_stack.t inside the denv if the current mutable
+   way proves to be too annoying.
+*)
+
+(* Individual decisions, with debuginfo *)
+type t = {
+  dbg : Debuginfo.t;
+  decision : Decision.t;
+}
+
+(* Actual log storage. During simplification, in order to be more efficient,
+   decisions are stored from the most recent one (in the head of the list),
+   to the oldest one (at the end of the list).
+   This means that one should rev the list before printing. *)
+let log : t list ref = ref []
+
+let stars fmt depth =
+  Format.fprintf fmt "%s" (String.make (depth + 1) '*')
+
+let rec print ~depth fmt = function
+  (* end of report log *)
+  | [] ->
+    if depth <> 0 then
+      Misc.fatal_errorf "incoherent depth at end of inlining report"
+
+  (* Entering a function declaration (possibly nested) *)
+  | { dbg; decision = At_function_declaration {
+      pass = Before_simplify; code_id; decision; } } :: r ->
+    Format.fprintf fmt "%a Definition of %s{%a}@\n"
+      stars depth (Code_id.name code_id) Debuginfo.print dbg;
+    Format.fprintf fmt "%a @[<v>Before simplification:@ @ %a@]@\n@\n"
+      stars (depth + 1)
+      Inlining_decision.Function_declaration_decision.report decision;
+    print ~depth:(depth + 1) fmt r
+
+  (* Function call *)
+  | { decision = At_call_site C_call; dbg; } :: r ->
+    Format.fprintf fmt "%a @[<v>%s of %s{%a}@ @ %s@ %s@]@\n@\n"
+      stars depth
+      (if depth = 0 then "Toplevel application" else "Application")
+      "<unknown function>" Debuginfo.print dbg
+      "The C function call has not been inlined"
+      "because the optimizer cannot inline C calls";
+    print ~depth fmt r
```

Comment:
I don't think these messages are particularly useful (C calls and method calls). I would skip reporting them completely.

Reply:
I've removed messages for C-calls and methods